docs/specs/getAgent.md

```diff
@@ -0,0 +1,146 @@
+
+**WindowProxy**: An interface defined by the HTML standard that proxies a remote Window object: https://html.spec.whatwg.org/multipage/nav-history-apis.html#the-windowproxy-exotic-object. This is used by WCP.
+
+**MessagePort**: An interface defined by the HTML Channel Messaging standard that can be used for cross-domain communication between windows. This is used by BCP.
+
+**Flux Standard Action (FSA)**: An [industry standard](https://github.com/redux-utilities/flux-standard-action) JSON envelope. All WCP and BCP messages are defined as FSAs.
+
+**SessionStorage**: A JavaScript storage API that persists data for web apps (segregated by domain) within a specific window or tab (SessionStorage is identical to the localStorage API but is tied to a particular window/tab, as defined by the HTML Standard: https://html.spec.whatwg.org/multipage/webstorage.html#dom-sessionstorage.)
+
+## Establishing Connectivity Using the Web Connection Protocol (WCP)
+
+The WCP algorithm (coordinated between `getAgent()` and DAs) has four steps. Each step may contain sub-steps.
+
+1. Connect to DA
+2. Validate app identity
+3. Persist DesktopAgentDetails to SessionStorage
+4. Resolve promise with DesktopAgent interface
+
+### Connect to DA (Step 1)
+
+1) Check for a possible navigation or refresh event
+
+    Check the SessionStorage key `FDC3-Desktop-Agent-Details` for a `DesktopAgentDetails` record. If it exists **and the appId matches**, then a navigation or refresh event has occurred and the stored data MUST be used to reestablish a connection to the DA. This ensures that the window maintains a consistent `instanceId` between navigation events within an app. If it doesn't exist, or the appId does not match, then proceed to step (2).
+
+    See [DesktopAgentDetails](./getAgent.d.ts).
+
+    Existing `DesktopAgentDetails` records MUST be used to limit discovery actions (in the next step) to the same mechanism as previously used or to skip the discovery step entirely if a `url` exists (immediately opening a hidden iframe).
+
+    If use of persisted data fails to establish a connection to the DA then `getAgent()` should reject its promise with "ReestablishConnectionFailed". 
+
+2) Create a race between these four mutually exclusive threads: 
+
+    **a) Timeout of 750ms** or the value provided by the `timeout` field
+
+    **b) Preload DAs:**
+
+    If a global (window or globalThis) `fdc3` object exists then return it immediately. If it does not exist then wait for the `fdc3Ready` event. After the `fdc3Ready` event triggers, recheck for the existence of the global `fdc3` object and return it if found.
+
+    **c) Browser-resident DAs:**
+
+    If a `url` exists from step (1) then proceed directly to step (iv), otherwise;
+
+    Locate all "Parents" (`window.opener`, `window.parent` or `window.parent.opener`). If no Parent exists then return null, otherwise simultaneously attempt to contact to each existing Parent with the following steps.
```

Comment:
See: https://github.com/finos-labs/fdc3-for-the-web/blob/8d26526a9dfe854c142221741a1be980d78fb985/packages/client/src/index.ts#L10

Not convinced we should have window.parent.opener as this might introduce some security issues.

Reply:
Could you elaborate on how window.parent.opener might introduce security issues we don't have with `window.parent` and `window.opener`? That wasn't immediately obvious to me.

We proposed this as it's necessary for platforms that want to wrap the opened apps in one of their own windows (e.g. to provide a channel selector, or multiple widgets in one window - I think that case was brought up by a couple of participants, so if we can support it securely we should try to do so).